orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java

```diff
@@ -0,0 +1,232 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.util.StringUtils;
+import org.springframework.web.client.HttpStatusCodeException;
+
+/** processes response or the client exception received from a webhook call. */
+public class WebhookResponseProcessor {
+
+  private final Logger log = LoggerFactory.getLogger(getClass());
+  private final ObjectMapper objectMapper;
+  private final WebhookStage.StageData stageData;
+  private final WebhookProperties webhookProperties;
+  private final StageExecution stageExecution;
+  private final String executionId;
+
+  public WebhookResponseProcessor(
+      ObjectMapper objectMapper,
+      StageExecution stageExecution,
+      WebhookProperties webhookProperties) {
+    this.objectMapper = objectMapper;
+    this.webhookProperties = webhookProperties;
+    this.stageExecution = stageExecution;
+    this.stageData = stageExecution.mapTo(WebhookStage.StageData.class);
+    this.executionId = stageExecution.getExecution().getId();
+  }
+
+  public TaskResult process(ResponseEntity<Object> response, Exception exceptionReceived) {
+    // Process exception if received.
+    if (exceptionReceived != null) {
+      if (exceptionReceived instanceof HttpStatusCodeException) {
+        return processReceivedHttpStatusException((HttpStatusCodeException) exceptionReceived);
+      } else {
+        return processClientException(exceptionReceived);
+      }
+    }
+
+    // Process received response.
+    if (response != null) {
+      return processResponse(response);
+    }
+
+    throw new IllegalArgumentException("No response or exception to process.");
+  }
+
+  TaskResult processReceivedHttpStatusException(HttpStatusCodeException e) {
+    Map<String, Object> stageOutput = initializeStageOutput();
+    Map<String, Object> webHookOutput = (Map<String, Object>) stageOutput.get("webhook");
+    webHookOutput.put("statusCode", e.getStatusCode());
+    webHookOutput.put("statusCodeValue", e.getStatusCode().value());
+    if (!StringUtils.isEmpty(e.getResponseBodyAsString())) {
+      webHookOutput.put("body", processResponseBodyAsJson(e.getResponseBodyAsString()));
+    }
+    TaskResult result = processReceivedFailureStatusCode(e.getStatusCode(), stageOutput);
+    log.warn(webHookOutput.get("error").toString(), e);
+    return result;
+  }
+
+  TaskResult processReceivedFailureStatusCode(HttpStatus status, Map<String, Object> stageOutput) {
+    String errorMessage;
+    ExecutionStatus executionStatus;
+    Map<String, Object> webHookOutput = (Map<String, Object>) stageOutput.get("webhook");
+    // Fail fast status check, retry status check or fail permanently
+    if (stageData.failFastStatusCodes != null
+        && stageData.failFastStatusCodes.contains(status.value())) {
+      errorMessage =
+          String.format(
+              "Received a status code %s configured to fail fast, terminating stage for pipeline %s to %s",
```

Comment:
While we're in here...

```suggestion
              "Received status code %s, which is configured to fail fast, terminating stage for pipeline %s to %s",
```